src/InteractionModel.H

```diff
@@ -41,80 +42,178 @@ namespace ExaEpi
+        + If the agent is #Status::immune, do nothing.
+        + If the agent is #Status::infected with the number of days infected (RealIdxDisease::disease_counter)
+          less than the incubation length, do nothing.
+        + Else if *i* is not infected and *j* is infected, compute probability of *i* getting infected
+          from *j* (see below).
+
+    Summary of how the probability of agent A getting infected from agent B is computed:
+    + Compute infection probability reduction factor from vaccine efficacy (#DiseaseParm::vac_eff)
+    + Within family - if their IntIdx::nborhood and IntIdx::family indices are same,
+      and the agents are at home:
+      + If B is a child, use the appropriate transmission probability (#DiseaseParm::xmit_child_SC or
+        #DiseaseParm::xmit_child) depending on whether B goes to school or not (#IntIdx::school)
+      + If B is an adult, use the appropriate transmission probability (#DiseaseParm::xmit_adult_SC or
+        #DiseaseParm::xmit_adult) depending on whether B works at a school or not (#IntIdx::school)
+*/
+template <typename AgentContainer, typename PTDType, typename CandidateFunc, typename BinaryInteractionFunc>
+void interactAgentsImpl(AgentContainer& agents, int bin_idx)
+{
+    BL_PROFILE("interactAgentsimpl");
+    int n_disease = agents.numDiseases();
+    IntVect bin_size = {AMREX_D_DECL(1, 1, 1)};
+    CandidateFunc isCandidate;
+    BinaryInteractionFunc binaryInteraction;
 
-                    AMREX_ALWAYS_ASSERT(np == bins_ptr->numItems());
+    for (int lev = 0; lev < agents.numLevels(); ++lev)
+    {
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+        for (MFIter mfi = agents.MakeMFIter(lev, TilingIfNotGPU()); mfi.isValid(); ++mfi) {
+            auto& ptile = agents.ParticlesAt(lev, mfi);
+            const auto& ptd = ptile.getParticleTileData();
+            const auto np = ptile.GetArrayOfStructs().numParticles();
+            auto& soa = ptile.GetStructOfArrays();
+            auto num_tiles = numTilesInBox(mfi.validbox(), true, bin_size);
+
+            int max_group = 0;
+            if (bin_idx != -1) {
+                auto bingroup_ptr = soa.GetIntData(bin_idx).data();
+                for (int i = 0; i < np; i++) {
+                    max_group = max(max_group, bingroup_ptr[i]);
                 }
-
+            }
+            max_group++;
+
+            // agents are binned by community + another group. These values can change from one time step to the next so they
+            // have to be recomputed (rebuild the bins) every time step
+            Binner<PTDType> binner(agents.Geom(lev), bin_size, mfi.validbox(), max_group, bin_idx);
+            DenseBins<PTDType> bins;
+            // The GPU bin policy is faster, but non-deterministic.
+            // bins.build(BinPolicy::GPU, np, ptd, num_tiles * max_group, binner);
+            bins.build(BinPolicy::Serial, np, ptd, num_tiles * max_group, binner);
```

Comment:
I think we should evaulate the performance in fast mode, i.e using the `BinPolicy::Gpu` method.